### ISSUE TITLE
Move pytest fixtures to ``conftest.py``

### ISSUE THREAD

**Issue by nstarman** [astropy/astropy#13886]
### Description

Pytest fixtures are best defined in a `conftest.py`. This has a number of advantages, not least of which is avoiding the annoying way that pytest is not compliant with F401 issues (See #13853).

### Additional context

https://docs.pytest.org/en/7.1.x/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files

**Comment by pllim** [comment]
A quick grep smoked out these files that should be investigated further (some might be false positive). Might need multiple PRs (one per subpackage) to get this done. Maybe a good thing to look at early in the v6 dev cycle (after the v5.2 branching and black stuff). Since `io.misc.asdf` is deprecated, I excluded it from the listing.

Also, is it ok to keep the subpackage-level fixtures in the subpackage-level conftest.py or is it cleaner to put everything in the astropy level conftest.py ?

* `astropy/coordinates/tests/test_earth_orientation.py`
* `astropy/coordinates/tests/test_iau_fullstack.py`
* `astropy/coordinates/tests/test_representation.py`
* `astropy/coordinates/tests/test_shape_manipulation.py`
* `astropy/coordinates/tests/test_representation_methods.py`
* `astropy/coordinates/tests/test_sky_coord_velocities.py`
* `astropy/coordinates/tests/test_spectral_coordinate.py`
* `astropy/cosmology/tests/test_core.py`
* `astropy/cosmology/tests/test_units.py`
* `astropy/cosmology/tests/test_parameter.py`
* `astropy/cosmology/tests/helper.py`
* `astropy/cosmology/tests/test_connect.py`
* `astropy/cosmology/flrw/tests/test_base.py`
* `astropy/cosmology/funcs/tests/test_comparison.py`
* `astropy/cosmology/io/tests/test_row.py`
* `astropy/cosmology/io/tests/test_cosmology.py`
* `astropy/cosmology/io/tests/test_mapping.py`
* `astropy/cosmology/io/tests/test_yaml.py`
* `astropy/cosmology/io/tests/test_model.py`
* `astropy/cosmology/io/tests/base.py`
* `astropy/cosmology/io/tests/test_table.py`
* `astropy/cosmology/io/tests/test_json.py`
* `astropy/cosmology/io/tests/test_ecsv.py`
* `astropy/io/ascii/tests/test_read.py`
* `astropy/io/ascii/tests/test_write.py`
* `astropy/io/ascii/tests/test_c_reader.py`
* `astropy/io/fits/tests/__init__.py`
* `astropy/io/fits/tests/test_image_dask.py`
* `astropy/io/misc/tests/test_hdf5.py`
* `astropy/io/registry/tests/test_registries.py`
* `astropy/io/votable/tests/table_test.py`
* `astropy/nddata/tests/test_ccddata.py`
* `astropy/samp/tests/test_client.py`
* `astropy/table/tests/test_bst.py`
* `astropy/table/tests/test_array.py`
* `astropy/table/tests/test_pprint.py`
* `astropy/table/tests/test_table.py`
* `astropy/table/tests/test_index.py`
* `astropy/tests/helper.py`
* `astropy/time/tests/test_custom_formats.py`
* `astropy/time/tests/test_precision.py`
* `astropy/timeseries/periodograms/bls/tests/test_bls.py`
* `astropy/timeseries/periodograms/lombscargle/tests/test_utils.py`
* `astropy/timeseries/periodograms/lombscargle/tests/test_lombscargle.py`
* `astropy/timeseries/periodograms/lombscargle/implementations/tests/test_utils.py`
* `astropy/timeseries/periodograms/lombscargle/implementations/tests/test_mle.py`
* `astropy/units/tests/test_quantity_decorator.py`
* `astropy/utils/tests/test_data.py`
* `astropy/visualization/wcsaxes/tests/test_wcsapi.py`
* `astropy/visualization/wcsaxes/tests/test_images.py`
* `astropy/visualization/wcsaxes/tests/test_coordinate_helpers.py`
* `astropy/visualization/wcsaxes/tests/test_misc.py`
* `astropy/wcs/tests/test_utils.py`
* `astropy/wcs/wcsapi/tests/test_fitswcs.py`
* `astropy/wcs/wcsapi/wrappers/tests/test_sliced_wcs.py`

**Comment by nstarman** [comment]
> A quick grep smoked out these files that should be investigated further (some might be false positive). Might need multiple PRs (one per subpackage) to get this done. Maybe a good thing to look at early in the v6 dev cycle (after the v5.2 branching and black stuff). Since `io.misc.asdf` is deprecated, I excluded it from the listing.

Wow. Thanks!

> Also, is it ok to keep the subpackage-level fixtures in the subpackage-level conftest.py or is it cleaner to put everything in the astropy level conftest.py ?

subpackage-level fixtures in the subpackage-level `conftest.py` are great!

**Comment by pllim** [comment]
So, we kinda have to be careful a little bit with this. I just ran into this problem with astropy v5.2rc1 testing downstream in `stsci.toos`.
 
```python
if ASTROPY_LT_5_2:
    from astropy.io.fits.tests import FitsTestCase
else:
    from astropy.io.fits.tests.conftest import FitsTestCase
```

**Comment by nstarman** [comment]
Maybe it's worth doing the rest in one PR and adding a whatsnew.

**Comment by Cadair** [comment]
> Pytest fixtures are best defined in a conftest.py

I disagree with this as a general policy. If a fixture is only used in a single test file, then I think it's much easier to read if you don't have to go and open a second file to find the fixture.

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/issues/13886#issuecomment-1344023631

@saimn , do you want that reverted for v5.2? Now is the time to do it before RC2.

**Comment by pllim** [comment]
FWIW, don't worry about the `stsci.tools` problem I mentioned above. I ended up copying the class entirely from v5.0.x branch downstream because of other issues like `setup` vs `setup_method`. For context, `stsci.tools` is the home for `stpyfits` that is quite hacky and there are no resources to refactor anything over there, so yeah...

I just brought it up as an example that someone could be using things we might move to `conftest` out there in the wild and this moving them will be a breaking change.

**Comment by nstarman** [comment]
> > Pytest fixtures are best defined in a conftest.py
> 
> 
> 
> I disagree with this as a general policy. If a fixture is only used in a single test file, then I think it's much easier to read if you don't have to go and open a second file to find the fixture.


It looks like ruff has a PR to support F811, so both flake8 and ruff will highlight the usage of pytest fixtures as F811 violations. The way Pytest is designed is very susceptible to this error, so pytest recommends putting fixtures in the conftest.py. We can choose not to and decorate our code with noqa:F811. Maybe in future there will be a plug-in to detect if a potential F811 is actually a pytest fixture and AOK.

**Comment by nstarman** [comment]
Oh. It looks like there is for flake8 and will be for pytest.
https://github.com/m-burst/flake8-pytest-style error PT019. I'll check later if pytest recommends this over our current implementation.

**Comment by saimn** [comment]
> @saimn , do you want that reverted for v5.2? Now is the time to do it before RC2.

No if it's ok for `stsci.tools` then no need to revert. I'm just a bit annoyed that a PR that adds "a few flake8 checks" ends with this kind of code change. And I don't think that `FitsTestCase` belongs in `conftest`, it's not a pytest thing.
`conftest.py` is a special file that acts as a plugin for pytest, e.g. allowing the fixtures to be used without importing them in test files. I don't think we should put things that don't need to be in this file.

**Comment by hamogu** [comment]
We discussed this [at the developer telecon](https://docs.google.com/document/d/15JSFh3OMF9Iz6ov3q_xxGO_BL8hRnuse4IMUrqEIvcg/edit#heading=h.y1b7rstf5zlm) and did not come of with any more arguments than above. Some consistency is good, but also don't want to break things for downstream users and don't necessarily need to create new conftest files for a thing that's used in just one test just for the sake of moving it.

In short, there is probably no on-size-fits-all answer to all the cases above - if somebody starts to do a few of those changes and opens a PR than we'll have to discuss the individual changes in that PR. That might be easier to do is smaller PRs.

**Comment by pllim** [comment]
I am clearing the milestone.
